sqlmesh/core/engine_adapter/clickhouse.py

```diff
@@ -0,0 +1,456 @@
+        temp_table = self._get_temp_table(target_table, **kwargs)
+
+        def query_factory() -> Query:
+            # It is possible for the factory to be called multiple times and if so then the temp table will already
+            # be created so we skip creating again. This means we are assuming the first call is the same result
+            # as later calls.
+            if not self.table_exists(temp_table):
+                self.create_table(
+                    temp_table, columns_to_types, storage_format=exp.var("MergeTree"), **kwargs
+                )
+
+                self.cursor.client.insert_df(temp_table.sql(dialect=self.dialect), df=df)
+
+            return exp.select(*self._casted_columns(columns_to_types)).from_(temp_table)
+
+        return [
+            SourceQuery(
+                query_factory=query_factory,
+                cleanup_func=lambda: self.drop_table(temp_table, **kwargs),
+            )
+        ]
+
+    def _get_data_objects(
+        self, schema_name: SchemaName, object_names: t.Optional[t.Set[str]] = None
+    ) -> t.List[DataObject]:
+        """
+        Returns all the data objects that exist in the given database.
+        """
+        query = (
+            exp.select(
+                exp.column("database").as_("schema_name"),
+                exp.column("name"),
+                exp.case(exp.column("engine"))
+                .when(
+                    exp.Literal.string("View"),
+                    exp.Literal.string("view"),
+                )
+                .else_(
+                    exp.Literal.string("table"),
+                )
+                .as_("type"),
+            )
+            .from_("system.tables")
+            .where(exp.column("database").eq(to_schema(schema_name).db))
+        )
+        if object_names:
+            query = query.where(exp.column("name").isin(*object_names))
+        df = self.fetchdf(query)
+        return [
+            DataObject(
+                catalog=None,
+                schema=row.schema_name,
+                name=row.name,
+                type=DataObjectType.from_str(row.type),  # type: ignore
+            )
+            for row in df.itertuples()
+        ]
+
+    def create_table_like(
+        self,
+        target_table_name: TableName,
+        source_table_name: TableName,
+        exists: bool = True,
+        **kwargs: t.Any,
+    ) -> None:
+        """
+        Create a table like another table or view.
+        """
+        target_table = exp.to_table(target_table_name)
+        source_table = exp.to_table(source_table_name)
+        on_cluster_sql = kwargs.get("ON_CLUSTER", None) or (
+            self.cluster if self.engine_run_mode.is_cluster else None
+        )
+        on_cluster_sql = (
+            f" ON CLUSTER {exp.to_identifier(on_cluster_sql)} " if on_cluster_sql else " "
+        )
+        create_sql = f"CREATE TABLE{' IF NOT EXISTS' if exists else ''} {target_table}{on_cluster_sql}AS {source_table}"
```

Comment:
Should this be done with sqlglot builders instead of strings?

Reply:
It's not currently parseable/generatable with sqlglot - i'll add it later this week and update